ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -260,31 +271,44 @@ object PackageService {
     ] => (Jwt, LedgerApiDomain.LedgerId) => Future[Set[TemplateId.RequiredPkg]]
 
   type ResolveChoiceArgType =
-    (TemplateId.RequiredPkg, Choice) => Error \/ iface.Type
+    (
+        ContractTypeId.Unknown.RequiredPkg,
+        Choice,
+    ) => Error \/ (Option[ContractTypeId.Interface.Resolved], iface.Type)
 
   type ResolveKeyType =
     TemplateId.RequiredPkg => Error \/ iface.Type
 
-  type ContractTypeIdMap = TemplateIdMap
-
-  case class TemplateIdMap(
-      all: Set[TemplateId.RequiredPkg],
-      unique: Map[TemplateId.NoPkg, TemplateId.RequiredPkg],
+  final case class ContractTypeIdMap[CtId[_]](
+      all: Set[ContractTypeId.Resolved[CtId[String]]],
+      unique: Map[CtId[Unit], ContractTypeId.Resolved[CtId[String]]],
   ) {
     // forms a monoid with Empty
-    def ++(o: TemplateIdMap): TemplateIdMap =
-      TemplateIdMap(all ++ o.all, (unique -- o.unique.keySet) ++ (o.unique -- unique.keySet))
+    def ++[O[X] >: CtId[X]](o: ContractTypeIdMap[O]): ContractTypeIdMap[O] = {
+      type UniqueGoal = Map[O[Unit], ContractTypeId.Resolved[CtId[String]]]
+      ContractTypeIdMap(
+        all ++ o.all,
+        ((unique.toMap: UniqueGoal) -- o.unique.keySet) ++ (o.unique -- unique.keySet),
```

Comment:
what is the advantage of using `keySet` over `keys` 

```suggestion
        ((unique.toMap: UniqueGoal) -- o.unique.keys) ++ (o.unique -- unique.keys)
```

Reply:
None, but there's no advantage to using `keys` either (its definition is `keySet`).

Reply:
fair enough.